### DISCUSSION TITLE
feature request: cancellable message generation, and re-generate

### DISCUSSION THREAD

**Discussion by wtlyu** [danny-avila/LibreChat#63] [Feature Requests & Suggestions]
have no idea on how to do that in server side, do you have any idea? @danny-avila

**Comment by danny-avila** [comment]
in express, I was reading that req can listen to the connection being closed/aborted, and perhaps throwing on close would do it. nodechatgpt-api doesn't have a timeout function but that would be a cool PR for that project

**Comment by wtlyu** [comment]
After read the code from node-chatgpt-api. it seems to be an option to do that.

https://github.com/waylaidwanderer/node-chatgpt-api/blob/main/src/ChatGPTClient.js#L357

```
            await this.getCompletion(
                payload,
                (progressMessage) => {
                    if (progressMessage === '[DONE]') {
                        return;
                    }
                    const token = this.isChatGptModel ? progressMessage.choices[0].delta.content : progressMessage.choices[0].text;
                    // first event's delta content is always undefined
                    if (!token) {
                        return;
                    }
                    if (this.options.debug) {
                        console.debug(token);
                    }
                    if (token === this.endToken) {
                        return;
                    }
                    opts.onProgress(token);
                    reply += token;
                },
                opts.abortController || new AbortController(),
            );
```

The `opts.abortController`

**Reply by danny-avila** [reply]
I see so you can pass your own abortcontroller to abort?

**Reply by wtlyu** [reply]
well, TBH I'm not very familiar with abortcontroller. It's some thing that can cancel a Promise I guess?

**Reply by wtlyu** [reply]
From Doc

> AbortController.abort()
> 
> The abort() method of the [AbortController](dfile:///Users/wentao.lyu/Library/Application%20Support/Dash/DocSets/JavaScript/JavaScript.docset/Contents/Resources/Documents/developer.mozilla.org/en-US/docs/Web/API/AbortController.html) interface aborts a DOM request before it has completed. This is able to abort [fetch requests](dfile:///Users/wentao.lyu/Library/Application%20Support/Dash/DocSets/JavaScript/JavaScript.docset/Contents/Resources/Documents/developer.mozilla.org/en-US/docs/Web/API/fetch.html), the consumption of any response bodies, or streams.

**Reply by danny-avila** [reply]
yup this would be how to do it! we can have the server listen to the request closing on the client side, and then use the abort method on the abortcontroller passed to the client option

**Comment by wtlyu** [comment]
BTW, in node-chatgpt-api, there is a title generate function, why did you write your own one? (but their editor do nothing on the language.)

**Reply by danny-avila** [reply]
There wasn't one before I wrote mine, I guess I could've PRed and still can

**Comment by wtlyu** [comment]
#80
